src/mlpack/methods/ann/layer/convolution_impl.hpp

```diff
@@ -386,53 +414,37 @@ void ConvolutionType<
     const size_t fullOutputOffset = offset * maps;
 
     // Iterate over input maps.
-    for (size_t inMap = 0; inMap < inMaps; ++inMap)
+    #pragma omp parallel for
+    for (size_t inMap = 0; inMap < (size_t) inMaps; ++inMap)
     {
       // Iterate over output maps.
+      MatType output;
       for (size_t outMap = 0; outMap < maps; ++outMap)
       {
-        MatType output;
-
         BackwardConvolutionRule::Convolution(
-            mappedError.slice(outMap + fullOutputOffset),
+            dilatedMappedError.slice(outMap + fullOutputOffset),
             rotatedFilters.slice((outMap * inMaps) + inMap),
             output,
-            strideHeight,
-            strideWidth);
-
-        // If the stride width or height is greater than 1, then we have to
-        // insert columns and rows into the convolution output.
-        if (strideWidth == 1 && strideHeight == 1)
-        {
-          if (usingPadding)
-          {
-            gTemp.slice(inMap + fullInputOffset) += output.submat(
-                padWLeft,
-                padHTop,
-                padWLeft + gTemp.n_rows - 1,
-                padHTop + gTemp.n_cols - 1);
-          }
-          else
-          {
-            gTemp.slice(inMap + fullInputOffset) += output;
-          }
-        }
-        else
-        {
-          // We must iterate over each element of the output and manually
-          // re-insert the stride.
-          size_t col = padWLeft;
-          for (size_t i = 0; i < output.n_cols; ++i)
-          {
-            size_t row = padHTop;
-            for (size_t j = 0; j < output.n_rows; ++j)
-            {
-              gTemp(row, col, inMap + fullInputOffset) += output(j, i);
-              row += strideHeight;
-            }
-            col += strideWidth;
-          }
-        }
```

Comment:
Okay, so I guess the bug was that I was inserting zeros for dilation *after* convolution, but it should instead be *before* convolution.  Correct me if I misunderstood. :+1: